### ISSUE TITLE
materialSwitch turning itself off

### ISSUE THREAD

**Issue by GreyEcologist** [CosmicMind/Material#195]
First I wanted to say, you guys are amazing! This library is insanely awesome. So thanks.

I'm using the materialSwitch inside UITableCells and every time I scroll a cell outside of the view its materialSwitch gets reset to its default off state. I implemented the materialSwitchStateChanged() delegate method inside my tableViewCell and it's being called every time I scroll. Shouldn't it only be called when the switch is actually being turned on or off?

**Comment by adamdahan** [comment]
Hey @GreyEcologist, thank you for the beautiful words. We really appreciate it. 

In regards to your issue: 

Can you share your cellForRowAtIndexPath method so we can take a deeper look? 

Immediately first thought that came to mind was that you weren't persisting the state of switch. But after reading your message a couple times I want to make sure your implementation is correct and if it is, something might be incorrect on our end.

Thanks!

**Comment by GreyEcologist** [comment]
Sure, I am using a asyncDisplayKit node 
```
final class TableViewCell: ASCellNode, MaterialSwitchDelegate { .. }
```
So the cellForRow equivalent is nodeforRow (I'm also using RAC):
```
    func tableView(tableView: ASTableView, nodeForRowAtIndexPath indexPath: NSIndexPath) -> ASCellNode {
        var node = ASCellNode()
        self.getStructSignal(index: indexPath.row)
            .on(next: { value in node = tableViewCell(struct: value) })
            .start()
        return node
    }
```
and the code below is inside my node init() method, and the switch itself is placed inside a subnode:
```
        let followSwitch = MaterialSwitch(size: .Small, state: self.isFollowed == true ? .On : .Off)
        followSwitch.center = CGPoint(x: Constants.kScreenWidth - 50, y: 32)
        followSwitch.buttonOnColor = Constants.kWineShade
        followSwitch.trackOnColor = followSwitch.trackOffColor
        self.switchNode = ASDisplayNode(viewBlock: { () -> UIView in return followSwitch })
        self.switchNode.preferredFrameSize = CGSize(width: 20, height: 20)

        super.init()
        followSwitch.delegate = self
        self.addSubnode(self.switchNode)
```
Everything works, just the materialSwitchStateChanged() is being called when I scroll.

**Comment by nhnam** [comment]
Hi GreyEcologist , As I remember , nodeforRow will be called by your tableview to get the next cell- which will be displayed on sreen. I see you create new node everytime as:

``` var node = ASCellNode()``
...
return node
`

should we use a reusable node here ? I also recommend caching value of switchers for each cell in your table view controller.

**Comment by danieldahan** [comment]
Hi @GreyEcologist, 

Your issue is interesting, because even though you are creating a new node each time, you are only setting the delegate after the creation of the node where you set the value. I am going to run a few tests and see if there is a regression issue with some of the latest updates we made. 

As a side note, you would see a gain in performance if you did not create a new ASCellNode each time, and then updated the respective UI elements. If you decide to go down that road, then I would suggest to remove the delegate when setting the MaterialSwitch, and then setting the delegate back, to avoid the call while updating the UI.

**Comment by GreyEcologist** [comment]
Hey guys, thanks for the help, I'm not an expert on async so I had to refresh my memory a bit but unlike regular table view cells the nodes don't implement reuse, the benefit is that they're initialized and laid out in a background thread before the view is laid out, and also the nodeForRow gets only called once and never again after that.
```
Instead of implementing -tableView:cellForRowAtIndexPath:, your data source must implement -tableView:nodeForRowAtIndexPath:. This method must be thread-safe and should not implement reuse. Unlike the UITableView version, it won't be called when the row is about to display.
```

It was working for me last week, so prolly version 1.34.x or something, I'm looking into it right now.

**Comment by GreyEcologist** [comment]
I can tell that 1.33.x was working for me, I'm trying to see which 1.34.x version stopped working for me, but cocoapods is only enabling me to get 1.34.10.

**Comment by GreyEcologist** [comment]
Ok, so the issue I'm running into starts with the changes introduced in 1.34.6. I'll look into the changes now.

**Comment by danieldahan** [comment]
Nice, I will take a look as well.

**Comment by GreyEcologist** [comment]
@danieldahan I can't tell from looking at the code, could be the changes in the setSwitchState(), but yeah, that's above my pay grade, lol. I do know the switch .ValueChanged behavior changed in 1.34.6. Hope that helps.

**Comment by danieldahan** [comment]
So 1.34.6 onwards is the same code. The issue must have come up from 1.34.5 - 1.34.6. 

Can you share the code that updates the state of the MaterialSwitch, because I am not 100% convinced this is a regression issue. As it seems to work everywhere, and the Async UI is operating on its rules.

**Comment by danieldahan** [comment]
I'll run more tests tonight and confirm it is or not Material :)

**Comment by GreyEcologist** [comment]
@danieldahan Hey Da I found it! The prepareView() had a gestureRecognizer, I couldn't see cuz it showed an empty function in the code compare but in the actual 1.34.5 code it had:

```
    public func prepareView() {
        addGestureRecognizer(UITapGestureRecognizer(target: self, action: "handleTapped:"))
    }
```

That function is also called to the two init() methods, when I add it to later version, my code works! Was it removed on purpose?

**Comment by danieldahan** [comment]
The UITapGestureRecognizer sets a handler for when the user taps the MaterialSwitch. I am curious as to how that solves the problem, as you wouldn't have the "handleTapped:" method implemented. Did you implement that method?

**Comment by GreyEcologist** [comment]
Hey, for me the gestureRecognizer calls the materialSwitchStateChanged() only when I switch it on/off. I could also add a target to the switch and listen for .ValueChanged events, either way work well for me when that handler is present, but neither work when it isn't there, as the .ValueChanged gets called whenever I scroll my tableview.

**Comment by danieldahan** [comment]
That is my concern. I cannot apply a fix, if one is needed, as a tap gesture makes no sense to influence a scroll. And I don't know if you implemented the handleTapped method. At this point, I still don't see a fix, or an issue. I will continue investigating it, and hopefully it will make more sense. All the best!

**Comment by GreyEcologist** [comment]
the handleTapped() was implemented in previous versions of the MaterialSwitch.swift:

```
/// Handles the tap gesture.
    internal func handleTapped(recognizer: UITapGestureRecognizer) {
        if true == CGRectContainsPoint(trackLayer.frame, layer.convertPoint(recognizer.locationInView(self), fromLayer: layer)) {
            setSwitchState(.On == internalSwitchState ? .Off : .On)
        }
    }
```

**Comment by danieldahan** [comment]
Hmm... The confusion is, you reported a scrolling problem that caused the delegate to fire, and now the tap gesture is what fixes it. I am trying to make sense of it, so I can fix the issue if there is one.

**Comment by GreyEcologist** [comment]
@danieldahan the issue is with the highlighted computed var: 
```
	public override var highlighted: Bool {
		get {
			return .On == internalSwitchState
		}
		set(value) {
			setOn(value, animated: true)
		}
	}
```

Currently the MaterialSwitch highlighted property gets called when the ASCellNode, inside of which I am using the switch, calls its own highlighted property. If I remove this var from the MaterialSwitch, it behaves just like a UISwitch.  Updating the internalSwitchState works as well:

```
	public override var highlighted: Bool {
		get {
			return .On == internalSwitchState
		}
		set(value) {
			//setOn(value, animated: true)
            internalSwitchState = value == true ? .On : .Off
		}
	}
```

**Comment by danieldahan** [comment]
So what should actually be happening, is the removal of the highlighted and selected properties. As those are internal for the UIControl to set. That should remove your issue. I am going to make a push with that, and if you can confirm it fixes it, then we are good :)

**Comment by GreyEcologist** [comment]
Cool! Sounds good.

**Comment by danieldahan** [comment]
Your fix is in version [1.35.1](https://github.com/CosmicMind/Material/releases/tag/1.35.1). If the issue persists, reopen the issue or create a new one. Thank you for submitting this :)

**Comment by GreyEcologist** [comment]
Works now! Thanks @danieldahan!
